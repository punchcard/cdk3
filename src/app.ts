/* eslint-disable prettier/prettier */
import * as path from "path";
import * as cdk from "@aws-cdk/core";
import * as cdk3 from ".";

// application for testing deployments
const app = new cdk.App();

const stack = new cdk.Stack(app, "test-cdk3-10", {
  env: {
    account: process.env.ENV_AWS_ACCOUNT_ID,
    region: process.env.ENV_AWS_REGION ?? "us-west-2",
  },
});

const wallet = new cdk3.Wallet(stack, "Wallet");

const testnet = new cdk3.LocalEthChain(stack, "Testnet");

new cdk3.Contract(stack, "HelloWorld", {
  owner: wallet,
  contractFile: path.join(__dirname, "..", "contracts", "my-erc20.sol"),
  chain: testnet,
});
