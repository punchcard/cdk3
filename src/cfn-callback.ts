import * as https from "https";
import * as url from "url";

import {
  CloudFormationCustomResourceEventCommon,
  CloudFormationCustomResourceResponse,
} from "aws-lambda";

/**
 * Success or Failure Response payloads of a CRUD operation..
 */
export type CloudFormationResponse = SuccessResponse | FailedResponse;

interface Response {
  PhysicalResourceId: CloudFormationCustomResourceResponse["PhysicalResourceId"];
  Data?: Record<string, any>;
}

/**
 * Successful CRUD operation.
 */
export interface SuccessResponse extends Response {
  Status: "SUCCESS";
}

/**
 * Failed CRUD operation.
 */
export interface FailedResponse extends Response {
  Status: "FAILED";
  /**
   * Reason why the operation failed.
   */
  Reason: string;
}

/**
 * Call back to CloudFormation.
 *
 * @param event CFN event we're responding to
 * @param body response body
 */
export async function callbackToCloudFormation(
  event: CloudFormationCustomResourceEventCommon,
  body: CloudFormationResponse
): Promise<void> {
  const payload = JSON.stringify({
    ...event,
    ...body,
  });
  console.log(payload);
  // TODO: utilize url.URL
  const parsedUrl = url.parse(event.ResponseURL);
  return new Promise<void>((resolve, reject) => {
    try {
      const request = https.request(
        {
          headers: {
            "content-type": "application/json",
            "content-length": payload.length,
          },
          hostname: parsedUrl.hostname,
          method: "PUT",
          path: parsedUrl.path,
          port: 443,
        },
        () => resolve()
      );
      request.on("error", reject);
      request.write(payload);
      request.end();
    } catch (err) {
      reject(err);
    }
  });
}
