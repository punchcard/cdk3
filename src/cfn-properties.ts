import { CloudFormationCustomResourceEventCommon } from "aws-lambda";

/**
 * Get a string from the ResourceProperties and throw an error if it does not exist or is not a string.
 *
 * @param properties resource properties - name value pairs from CloudFormation
 * @param name name of the property to get
 */
export function getString(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): string {
  const prop = getStringOrUndefined(event, name);
  if (prop === undefined) {
    throw new Error(
      `expected a string resource property, '${name}', but got '${prop}'`
    );
  }
  return prop;
}

export function getStringOrUndefined(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): string {
  const prop = event.ResourceProperties[name];
  if (prop === undefined || typeof prop === "string") {
    return prop;
  }
  throw new Error(
    `expected a string resource property, '${name}', but got '${prop}'`
  );
}

export function getNumberOrUndefined(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): number | undefined {
  if (name in event.ResourceProperties) {
    return getNumber(event, name);
  } else {
    return undefined;
  }
}

export function getNumber(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): number {
  const prop = event.ResourceProperties[name];
  if (typeof prop === "number") {
    return prop;
  } else if (typeof prop === "string") {
    try {
      return parseInt(prop, 10);
    } catch (err) {
      try {
        return parseFloat(prop);
      } catch {
        // no-op
      }
    }
  }
  throw new Error(
    `expected a number resource property, '${name}', but got '${prop}'`
  );
}
