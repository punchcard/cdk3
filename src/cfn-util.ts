import * as https from "https";
import * as url from "url";

import {
  CloudFormationCustomResourceEventCommon,
  CloudFormationCustomResourceResponse,
} from "aws-lambda";

/**
 * Success or Failure Response payloads of a CRUD operation..
 */
export type CloudFormationResponse = SuccessResponse | FailedResponse;

interface Response {
  PhysicalResourceId: CloudFormationCustomResourceResponse["PhysicalResourceId"];
  Data?: Record<string, any>;
}

/**
 * Successful CRUD operation.
 */
export interface SuccessResponse extends Response {
  Status: "SUCCESS";
}

/**
 * Failed CRUD operation.
 */
export interface FailedResponse extends Response {
  Status: "FAILED";
  /**
   * Reason why the operation failed.
   */
  Reason: string;
}

/**
 * Call back to CloudFormation.
 *
 * @param event CFN event we're responding to
 * @param body response body
 */
export async function callbackToCloudFormation(
  event: CloudFormationCustomResourceEventCommon,
  body: CloudFormationResponse
): Promise<void> {
  const payload = JSON.stringify({
    ...event,
    ...body,
  });
  console.log(payload);
  // TODO: utilize url.URL
  const parsedUrl = url.parse(event.ResponseURL);
  return new Promise<void>((resolve, reject) => {
    try {
      const request = https.request(
        {
          headers: {
            "content-type": "application/json",
            "content-length": payload.length,
          },
          hostname: parsedUrl.hostname,
          method: "PUT",
          path: parsedUrl.path,
          port: 443,
        },
        () => resolve()
      );
      request.on("error", reject);
      request.write(payload);
      request.end();
    } catch (err) {
      reject(err);
    }
  });
}

/**
 * Get a string from the ResourceProperties and throw an error if it does not exist or is not a string.
 *
 * @param properties resource properties - name value pairs from CloudFormation
 * @param name name of the property to get
 */
export function getString(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): string {
  const prop = getStringOrUndefined(event, name);
  if (prop === undefined) {
    throw new Error(
      `expected a string resource property, '${name}', but got '${prop}'`
    );
  }
  return prop;
}

export function getStringOrUndefined(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): string {
  const prop = event.ResourceProperties[name];
  if (prop === undefined || typeof prop === "string") {
    return prop;
  }
  throw new Error(
    `expected a string resource property, '${name}', but got '${prop}'`
  );
}

export function getNumberOrUndefined(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): number | undefined {
  if (name in event.ResourceProperties) {
    return getNumber(event, name);
  } else {
    return undefined;
  }
}

export function getNumber(
  event: CloudFormationCustomResourceEventCommon,
  name: string
): number {
  const prop = event.ResourceProperties[name];
  if (typeof prop === "number") {
    return prop;
  } else if (typeof prop === "string") {
    try {
      return parseInt(prop, 10);
    } catch (err) {
      try {
        return parseFloat(prop);
      } catch {
        // no-op
      }
    }
  }
  throw new Error(
    `expected a number resource property, '${name}', but got '${prop}'`
  );
}
