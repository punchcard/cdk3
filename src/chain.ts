import * as ec2 from "@aws-cdk/aws-ec2";
import * as ecs from "@aws-cdk/aws-ecs";
import * as elb from "@aws-cdk/aws-elasticloadbalancingv2";
import * as cdk from "@aws-cdk/core";
import { TestMnemonic } from ".";

export interface IChain {
  readonly chainName: string;
  readonly chainId: number;
  readonly rpcUrl?: string;
}

export class Chain implements IChain {
  /**
   * @see https://ethereum.org/
   */
  public static readonly Ethereum = new Chain("Ethereum", 1, "ETH");

  /**
   * @see https://www.binance.org/en/smartChain
   */
  public static readonly BinanceSmartChain = new Chain(
    "Binance Smart Chain",
    56,
    "BNB",
    "https://bsc-dataseed.binance.org/",
    "https://bscscan.com"
  );

  /**
   * @see https://www.avax.network/
   */
  public static readonly Avalance = new Chain(
    "Avalanche",
    43114,
    "AVAX",
    "https://api.avax.network/ext/bc/C/rpc",
    "https://snowtrace.io/"
  );

  /**
   * @see https://fantom.foundation/
   */
  public static readonly Fantom = new Chain(
    "Fantom",
    250,
    "FTM",
    "https://rpc.ftm.tools/",
    "https://ftmscan.com/"
  );

  constructor(
    readonly chainName: string,
    readonly chainId: number,
    readonly symbol: string,
    readonly rpcUrl?: string,
    readonly blockExplorer?: string
  ) {}
}

export interface LocalEthChainProps {
  /**
   * VPC to run the local network win.
   *
   * @default - a VPC is created for you.
   */
  readonly vpc?: ec2.Vpc;
}

export function isTestChain(chain: IChain): chain is TestChain {
  return (chain as TestChain).kind === "TestChain";
}

/**
 * Creates a local Ethereum chain hosted on an ECS Fargate cluster.
 */
export class TestChain extends cdk.Construct implements IChain {
  readonly kind: "TestChain" = "TestChain";

  readonly chainName: string;
  readonly chainId: number;
  readonly rpcUrl?: string;

  readonly vpc: ec2.Vpc;
  readonly service: ecs.FargateService;
  readonly cluster: ecs.Cluster;
  readonly task: ecs.TaskDefinition;
  readonly loadBalancer: elb.NetworkLoadBalancer;

  constructor(
    scope: cdk.Construct,
    id: string,
    props: LocalEthChainProps = {}
  ) {
    super(scope, id);

    this.chainName = "Ethereum";
    this.chainId = 1;
    this.vpc = props.vpc ?? new ec2.Vpc(this, "VPC");

    this.cluster = new ecs.Cluster(this, "Cluster", {
      vpc: this.vpc,
    });
    this.task = new ecs.TaskDefinition(this, "Node", {
      compatibility: ecs.Compatibility.FARGATE,
      cpu: "256",
      memoryMiB: "512",
    });
    this.task.addContainer("Ganache", {
      image: ecs.ContainerImage.fromRegistry("trufflesuite/ganache-cli"),
      command: ["-m", `"${TestMnemonic}"`],
      portMappings: [
        {
          containerPort: 8545,
          hostPort: 8545,
        },
      ],
    });
    this.service = new ecs.FargateService(this, "Service", {
      cluster: this.cluster,
      taskDefinition: this.task,
    });

    this.loadBalancer = new elb.NetworkLoadBalancer(this, "LoadBalancer", {
      vpc: this.vpc,
      internetFacing: false,
    });
    const listener = this.loadBalancer.addListener("NodeListener", {
      port: 8545,
    });
    listener.addTargets("ECS", {
      port: 8545,
      targets: [this.service],
    });

    this.rpcUrl = `https://${this.loadBalancer.loadBalancerDnsName}`;
  }
}
