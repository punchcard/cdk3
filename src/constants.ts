/**
 * Names of environment variable keys used to share information from CDK and Lambda.
 */
export enum EnvironmentKeys {
  Address = "Address",
  ChecksumAddress = "ChecksumAddress",
  PublicKey = "PublicKey",
  WalletName = "WalletName",
  WalletSecretArn = "WalletSecretArn",
}
