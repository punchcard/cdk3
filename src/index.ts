export * from "./chain";
export * from "./contract";
export * from "./test-chain";
export * from "./test-keys";
export * from "./wallet";
