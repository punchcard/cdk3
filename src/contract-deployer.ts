import { CloudFormationCustomResourceEvent } from "aws-lambda";
import * as AWS from "aws-sdk";

import * as ethers from "ethers";
import { callbackToCloudFormation, getString } from "./cfn-util";
import { EnvironmentKeys } from "./constants";

const s3 = new AWS.S3();

/**
 * Entrypoint to the Contract Deployer Lambda Function. This Function is called
 * as part of the CloudFormation CRUD lifecycle for the Custom::Contract Resource.
 */
export async function handle(event: CloudFormationCustomResourceEvent) {
  try {
    const bucketName = getString(event, EnvironmentKeys.ContractBucketName);
    const objectKey = getString(event, EnvironmentKeys.ContractObjectKey);

    const contractObject = await s3
      .getObject({
        Bucket: bucketName,
        Key: objectKey,
      })
      .promise();

    if (event.RequestType === "Create") {
      // TODO
      const contractFactory = ethers.ContractFactory.fromSolidity(
        contractObject.Body
      );
    }
  } catch (err) {
    await callbackToCloudFormation(event, {
      Status: "FAILED",
      PhysicalResourceId:
        event.RequestType === "Create" ? "" : event.PhysicalResourceId,
      Reason: (err as any).message,
    });
  }
}

// const factory = new ContractFactory(contractAbi, contractByteCode);

// // If your contract requires constructor args, you can specify them here
// const contract = await factory.deploy(deployArgs);

// console.log(contract.address);
// console.log(contract.deployTransaction);
