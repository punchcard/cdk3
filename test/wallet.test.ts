import "jest";
import * as cdk3 from "../src";
import { stackTest } from "./util";

test(
  "default wallet should provision an AWS Secret and KMS Encryption Key",
  stackTest((stack) => {
    new cdk3.Wallet(stack, "Wallet");
  })
);
