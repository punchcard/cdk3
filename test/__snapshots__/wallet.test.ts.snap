// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default wallet should provision an AWS Secret and KMS Encryption Key 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5bArtifactHash16EA8C07": Object {
      "Description": "Artifact hash for asset \\"8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5b\\"",
      "Type": "String",
    },
    "AssetParameters8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5bS3Bucket4FD5CEB7": Object {
      "Description": "S3 bucket for asset \\"8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5b\\"",
      "Type": "String",
    },
    "AssetParameters8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5bS3VersionKey0203572F": Object {
      "Description": "S3 key for asset version \\"8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5b\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "WalletEncryptionKey8455325D": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "Description": "Encryption Key securing the Wallet.",
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Condition": Object {
                "StringEquals": Object {
                  "kms:ViaService": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "secretsmanager.",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ".amazonaws.com",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:CreateGrant",
                "kms:DescribeKey",
              ],
              "Condition": Object {
                "StringEquals": Object {
                  "kms:ViaService": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "secretsmanager.",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ".amazonaws.com",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "WalletWalletResourceHandlerServiceRole65EB541B",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Condition": Object {
                "StringEquals": Object {
                  "kms:ViaService": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "secretsmanager.",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ".amazonaws.com",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "WalletWalletResourceHandlerServiceRole65EB541B",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "WalletPrivateKey1C295680": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "GenerateSecretString": Object {},
        "KmsKeyId": Object {
          "Fn::GetAtt": Array [
            "WalletEncryptionKey8455325D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "WalletWalletResourceHandler2D2301E3": Object {
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5bS3Bucket4FD5CEB7",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5bS3VersionKey0203572F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8857b364800a1b6cd0dbcde01dbb4440826030bf852a0d2669f521e8363e4f5bS3VersionKey0203572F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "SecretArn": Object {
              "Ref": "WalletPrivateKey1C295680",
            },
          },
        },
        "Handler": "index.handle",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "WalletWalletResourceHandlerServiceRole65EB541B",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WalletWalletResourceHandlerServiceRole65EB541B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WalletWalletResourceHandlerServiceRoleDefaultPolicyB0C4B71A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "WalletEncryptionKey8455325D",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "secretsmanager:PutSecretValue",
                "secretsmanager:UpdateSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "WalletPrivateKey1C295680",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WalletWalletResourceHandlerServiceRoleDefaultPolicyB0C4B71A",
        "Roles": Array [
          Object {
            "Ref": "WalletWalletResourceHandlerServiceRole65EB541B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
